### PR TITLE
feat(#715): asymmetric CTMRG root implicit AD — forward + root (Phase 1, partial)

### PR DESCRIPTION
**Pull request by yingjerkao**
> 🤖 **AI-generated PR** — written by Claude Code, posted by @yingjerkao.

Phase 1 of #715, **partial**. Stacked on #716 (Phase 0). §V of [arXiv:2607.15030](https://arxiv.org/abs/2607.15030): asymmetric CTMRG, 1×1 unit cell, four distinct corners and edges.

## Verified

| claim | measurement |
|---|---|
| Fishman half-infinite projectors close | `P_bot @ P_top ∝ 1` to 1e-15 |
| Forward energy is right | matches plain sweep iteration to 1e-13; that energy's explicit-backprop gradient matches symmetric FD to **10 digits** |
| All 20 characteristic equations (Eqs. 76–80) hold at the root | ‖F(y*)‖ = 2.5e-16 |
| `∂_yF` is invertible — implicit function theorem applies | cond ≈ 8e3, no zero mode |
| `F` is contractions only | no `svd`/`eigh` in the jaxpr |

## Three things the paper leaves implicit

Each cost a debugging cycle; all three are documented in the module:

1. **The sweep must be simultaneous, not Gauss–Seidel.** Eqs. 76–77 evaluate all four moves at the same `y`, so a sequential sweep's fixed point is not a root.
2. **Eq. 68 projects each corner on *both* open legs**, using projectors from two different moves.
3. **The per-bond phase must be pinned or nothing converges element-wise.** Corner spectra are invariant under bond rotations, so they converge to 1e-14 while signs keep flipping — and `F` compares tensors, not magnitudes. Pinning the phase on `U` alone is insufficient (the projectors inherit the previous bond gauge through the quadrants); it has to be pinned on `P_top`, and near-degenerate singular values make an `argmax` rule oscillate with period two, so it aligns to the previous sweep instead.

## Not verified — the gradient

This module keeps `S` diagonal and uses the standard Fishman projectors, **skipping the modified corners/edges of Eqs. 73–75/82**. I had documented that as a stability-only shortcut. It is not:

> implicit **−1.7898** vs explicit-backprop reference **−0.81738127494**

with `‖F(y*)‖ = 1e-16` and a well-conditioned Jacobian. The `s^-1/2` inside the projectors breaks covariance under the eight environment gauge unitaries (Eq. 86), so restricting the isometry variation to its null space (Eq. 88) discards a *real* contribution rather than a gauge one.

**So §V.3's modified variables are load-bearing for correctness, not just for stability under degenerate singular values.** That is the main finding of this PR and it changes the Phase 1 estimate in #715.

`test_gradient_parity_needs_the_modified_variables` is a **strict xfail** carrying that measurement; completing Eqs. 73–75/82 flips it green.

## Merge guidance

Reviewable and safe to land as infrastructure (nothing existing is touched, no public API), or hold until the gradient is finished — your call. Phases 2 and 3 are blocked on this, since both build on the §V equations.

7 passing tests + 1 documented xfail, `algorithm` bucket, 27 s.

### CODE REVIEW

**Comment by yingjerkao** [comment]
> 🤖 **AI-generated comment** — written by Claude Code, posted by @yingjerkao.

Added `docs/plans/2026-07-29-715-phase1-modified-variables.md` (ff460de) with the construction the gradient failure points to, written in this module's conventions rather than the paper's diagrams — Eqs. 73–82 are images, so the leg wiring had to be re-derived.

Short version:

- `P_top = P̃_top s^-1/2`, `P_bot = s^-1/2 P̃_bot` with `P̃_top = bot Vh†`, `P̃_bot = U† top`, so a bond insertion becomes `P̃_top s^-1 P̃_bot`. `s^-1` transforms covariantly where `s^-1/2` does not — that single change is what buys covariance.
- Substituting back gives `C = s_a^-1/2 C̃ s_b^-1/2` (and likewise for edges), which is Eq. 82; the shifted index in Appendix F Table 1 is just the neighbouring move.
- `S` must become a general χ×χ matrix — with `U`, `V` restricted to null-space variations the in-space rotation has nowhere else to go. Exact analogue of Phase 0's "treat `C` as generic complex Hermitian in the reverse pass".
- The energy still consumes `x = (C, E)`, so `ў` routes through the `y ↦ x` map (Algorithm 2 line 7).

Two things worth flagging for whoever implements it:

**There is no third option.** Promoting the in-space rotation `δω` to a variable does not work: the equation that would determine it is the diagonality of `S`, and linearising that *is* the `1/(s_i² − s_j²)` divergence the method exists to avoid. Either the equations are covariant, or the divergence comes back.

**Do not use `eigh` for the matrix roots** (`s^-1/2`, `s^L = (s†s)^1/4`, `s^R`). The composite is smooth at degeneracies — a Löwner divided difference tends to `f'(λ)` — but JAX's `eigh` VJP still divides by `λ_i − λ_j` and will NaN, which would reintroduce the exact failure mode the paper removes. Newton–Schulz / Denman–Beavers is matrix multiplications only, so `F` stays decomposition-free and `test_no_svd_in_the_differentiated_equations` keeps passing.

The doc also lists the work order and the oracles to check at each step. `‖F(y*)‖` is the sharp one — it went 3.3e3 → 0.24 → 2.5e-16 as the conventions were fixed the first time, and it localises an error to a single equation.

**Comment by yingjerkao** [comment]
> 🤖 **AI-generated comment** — written by Claude Code, posted by @yingjerkao.

Pushed 84f05f1. The gradient gap is now mostly closed, and the cause turned out to be smaller than the full §V.3 rewrite I described above.

**Gradient error vs explicit backprop** (which matches symmetric FD to ten digits), four states, two χ:

| | error |
|---|---|
| before | 1.2e0 |
| after | **1.1e-2 … 2.5e-2** |

Root stays at machine precision (2.5e-16) throughout.

## What it was

`S` being a *diagonal vector*. Under the bond rotation `U → U W`, `Vh → W† Vh`, `S → W† S W`, a genuine **matrix** inverse square root is equivariant — `(W† S W)^-1/2 = W† S^-1/2 W` — so `P_bot @ P_top = S^-1/2 (U† M Vh†) S^-1/2` survives for any `S`. With `S` diagonal the in-space rotation is not representable at all, the closure breaks at first order, and Eq. 88's null-space restriction then discards a *physical* contribution instead of a gauge one.

Promoting `S` to a general matrix and Eq. 79 to the full χ×χ block is exactly what the paper means by "variations of the singular value matrices become general non-diagonal complex matrices" — the analogue of Phase 0's "treat `C` as a generic complex Hermitian matrix in the reverse pass".

The root is **Denman–Beavers, not `eigh`**. `test_no_svd_in_the_differentiated_equations` earned its keep here: it caught an `svd` I had briefly introduced *just to compute a scale factor*, which would have quietly put a decomposition back into the gradient path.

## Two dead ends, recorded so they aren't retried

- **Eq. 73's quartic roots inside the projectors make it worse** (2.0e-1). They are equivariant under independent left/right rotations, but their product is not `S^-1`, so the closure breaks at first order in a non-diagonal `S`. In the paper they sit on the *cut legs* of the modified environment, where no closure condition applies.
- **Rescaling Denman–Beavers by the extreme singular values does not fix the χ=6 NaN.** The cause is a rank-deficient half-infinite environment in early sweeps; flooring the retained spectrum fixes it.

## Still open

The residual 1–2.5% needs the rest of Eqs. 73–82 — modified corners/edges carrying `s` explicitly on the bonds, plus the cut-leg roots. It is stable under the reference's sweep count (12 vs 24 sweeps: identical to four digits), so it is a structural term, not convergence noise. The design note is updated and the xfail carries the current number.

**Comment by yingjerkao** [comment]
> 🤖 **AI-generated comment** — written by Claude Code, posted by @yingjerkao.

The remaining gradient gap is now tracked separately as #718 — freezing `U*`, `Vh*` inside `F` drops the `∂_cF · dc/dp` term, measured directly as `dE/dU* ≈ 1.5e-2` against a missing `7.14e-3`.

Everything else in this PR is verified: equations and root at 2.5e-16, the `y → x` map to 12 digits against the energy, the solve to 11 digits against a dense reference, cond 3.6e2 after equilibration with no null space, and reparametrisation invariance demonstrated.

`test_gradient_parity_needs_the_modified_variables` stays a strict xfail carrying the number; it flips green when #718 is resolved.

**Comment by yingjerkao** [comment]
> 🤖 **AI-generated comment** — written by Claude Code, posted by @yingjerkao.

Both findings addressed in 4337143.

**Withhold the known-invalid gradient.** Agreed. `asym_root_implicit_energy_and_grad`
now raises `NotImplementedError` referencing #718 unless the caller passes
`allow_known_invalid_gradient=True`, so a ~2.5%-wrong gradient can't reach an
optimiser by accident. The forward energy is unaffected and stays usable.

**Quadrant wiring.** The specific claim — that `C1.d` is adjacent to `T4.d` — is
not correct, and I checked it against the library's own conventions rather than
against my comments. `initialize_ctm_tensor_env` gives:

```
C1: ['c1_d', 'c1_r']
T4: ['t4_d', 'l2', 't4_u']
```

`T4` sits *below* `C1`, so the corner's downward leg meets the edge's **upward**
leg: `c1_d ↔ t4_u`. The einsum `"ce,efg,hic,fjik->gkhj"` puts `c` on `T4`'s index
2, which is `t4_u`. `_lower_left_quadrant` checks out the same way
(`c4_r↔t3_l`, `c4_u↔t4_d`, a-legs `d2`/`l2`).

**But the underlying point stands and was worth acting on.** Every endpoint in a
quadrant has dimension `chi` or `D**2`, so a transposed contraction is
shape-legal and silently wrong, and no existing test could distinguish one.
`test_quadrant_wiring_matches_the_library_leg_labels` now rebuilds both quadrants
from the adjacency the leg *names* encode (`*_d` meets `*_u`, `*_r` meets `*_l`,
an edge's double-layer leg meets the matching leg of `a`) and pins the
hand-written einsums against it — 2e-16 agreement. It also asserts the
transposed variant from the review differs (3.4e-2), so the test can fail.

One wrinkle worth recording: the test only bites on a **converged** environment.
The initial one is up/down symmetric, and on it the two wirings agree *exactly*
— my first version of this check passed while comparing two different
contractions.

**A correction to something I reported earlier in this PR.** While chasing the
wiring claim I compared the asymmetric path against production `ctm_tensor` and
found gaps of 1e-3…1e-1 that looked like a forward bug. That inference was
wrong, and I want it on the record rather than buried: production's default
`svd` projector collapses to a **rank-1 corner** on the near-product and random
tensors I was testing, and production-`svd` vs production-`eigh` differ by 0.12
on the same state with neither χ-converged. Production disagrees with itself
there, so it isn't a usable oracle in that regime. On every well-conditioned
state — including a simple-update state that converged to the classical Néel
point — the asymmetric path agrees with production to ~1e-16. **There is no
evidence of a wiring or forward bug**; the earlier numbers were an artifact of
the test states. The `_ctm_tensor_energy` energy function is shared, so it was
never a candidate.

File: src/tenax/algorithms/_ctm_root_implicit_asym.py [hunk 1]
```diff
@@ -0,0 +1,649 @@
...
+    """``C1·T1·T4·a`` with axes ``(chi_r, a_r, chi_d, a_d)``.
+
+    The ``(chi_d, a_d)`` pair is the vertical bond to be truncated; the
+    ``(chi_r, a_r)`` pair is what remains open to the right.
+    """
+    # C1[c,e] T1[e,f,g] T4[h,i,c] a[f,j,i,k]
+    #   c = C1.d = T4.u   e = C1.r = T1.l   f = T1's a-leg = a.u
+    #   g = T1.r          h = T4.d          i = T4's a-leg = a.l
+    #   j = a.d           k = a.r
+    return jnp.einsum("ce,efg,hic,fjik->gkhj", env.C1, env.T1, env.T4, a)
```
[the remark this answers is not in the document]
Comment (yingjerkao) [reply src/tenax/algorithms/_ctm_root_implicit_asym.py:147]: > 🤖 **AI-generated comment** — written by Claude Code, posted by @yingjerkao.

Carried forward to **#800** rather than closed, because I could not cheaply confirm it either way on current `main`: `rotate_env` permutes the whole corner/edge set before the quadrant contraction, so the raw einsum indices no longer read off directly against the declared layouts.

Flagging rather than dropping precisely because of the failure mode you describe — a reversed endpoint pairing where every endpoint has dimension χ, so shape and fixed-point tests cannot see it. That is the same shape as #700 and #702, both of which were real and both of which converged to *a* fixed point while being wrong. It deserves one careful pass by someone with the Appendix-F index tables open.

[the remark this answers is not in the document]
Comment (yingjerkao) [reply src/tenax/algorithms/_ctm_root_implicit_asym.py:147]: > 🤖 **AI-generated comment** — written by Claude Code, posted by @yingjerkao.

Verified **resolved** on `main`: the strict-xfail this cited is gone — neither `1.7898` nor `0.81738127494` appears anywhere in `tests/` any more. The gradient was wrong for the reason later diagnosed in #718, and the covariant characteristic equations that fixed it landed across #719/#720/#722. `main` now carries a production test that runs a physical state through `optimize_gs_ad` on this engine and asserts descent plus a variational bound, rather than an expected failure. The "withhold it until it is right" request was answered by making it right.
